### PR TITLE
bpf map context and explicit pin path removed

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

    [BPF] no explicit pin file name in maps
    
    Makes easier to change the location of the pins and easier to maintain
    maps and add new ones.
    
    Allow multiple char version.

    [BPF] remove MapContext

    [BPF] remove RepinningEnabled from MapContext
    
    It is pretty much a global setting, mostly turned on.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf/bpfmap/bpf_maps.go [hunk 1]
```diff
@@ -34,101 +32,99 @@ import (
...
+		m.ArpMap,
+		m.FailsafesMap,
+		m.FrontendMap,
+		m.BackendMap,
+		m.AffinityMap,
+		m.RouteMap,
+		m.CtMap,
+		m.SrMsgMap,
+		m.CtNatsMap,
 	}
```
Comment (StevenTigera) [inline felix/bpf/bpfmap/bpf_maps.go:64]: QQ: for symmetry did we want/need to include `IfStateMap` and `RuleCountersMap` ?

File: felix/bpf/maps.go [hunk 2]
```diff
@@ -380,12 +423,21 @@ func (b *PinnedMap) Delete(k []byte) error {
 }
 
 func (b *PinnedMap) updateDeltaEntries() error {
+	logrus.WithField("name", b.Name).Debug("updateDeltaEntries")
+
+	if b.oldfd == b.fd {
+		return fmt.Errorf("old and new maps are the same")
+	}
```
Comment (StevenTigera) [inline felix/bpf/maps.go:430]: QQ: for completeness would it help / add value to add the FD similar to the log entry in L 423?

File: felix/bpf/maps.go [hunk 3]
```diff
@@ -708,19 +769,23 @@ func (b *PinnedMap) CopyDeltaFromOldMap() error {
...
+		if len(fname) >= len(name) && fname[0:len(name)] == name {
+			oldIdx := strings.Index(fname, "_old")
+			if oldIdx == -1 {
+				oldIdx = len(fname)
+			}
+			oldVersion, err = strconv.Atoi(fname[len(name):oldIdx])
 			if err != nil {
-				return 0, fmt.Errorf("invalid version %w", err)
+				// We may have names that have the same prefix. Don't error,
+				// just continue. We eventually run out of maps.
```
Comment (StevenTigera) [inline felix/bpf/maps.go:787]: nit: "we will eventually"

File: felix/bpf/counters/map.go [hunk 4]
```diff
@@ -17,34 +17,30 @@ package counters
...
 const PolicyMapKeySize = 8
 const PolicyMapValueSize = 8
 
 var MapParameters = bpf.MapParameters{
-	Filename:   "/sys/fs/bpf/tc/globals/cali_counters",
 	Type:       "percpu_array",
 	KeySize:    counterMapKeySize, // __u32
 	ValueSize:  counterMapValueSize * MaxCounterNumber,
 	MaxEntries: 1,
 	Name:       bpf.CountersMapName(),
```
Comment (sridhartigera) [inline felix/bpf/counters/map.go:31]: Should we make it similar to other maps by specifying the Name, Version and do away with ```bpf.CounterMapName```?

**Review by sridhartigera — APPROVED** [review]
LGTM.

File: felix/bpf/counters/map.go (hunk 4 above)
Comment (tomastigera) [reply felix/bpf/counters/map.go:31]: yes, in another PR that I am working on. Counters are sort of special, per iface and there is some workarounds that need this, so I am leaving it here for now. Jump maps have the same.

File: felix/bpf/bpfmap/bpf_maps.go (hunk 1 above)
Comment (tomastigera) [reply felix/bpf/bpfmap/bpf_maps.go:64]: not now, i want to keep it the way it used to be, it is unrelated.

**Pull request merged**